s3fs/core.py

```diff
@@ -967,18 +967,37 @@ async def _ls(self, path, detail=False, refresh=False, versions=False):
                 return files
         return files if detail else sorted([o["name"] for o in files])
 
+    def _exists_from_cache(self, path, bucket, key, version_id):
+        fullpath = "/".join((bucket, key))
+
+        try:
+            entries = self._ls_from_cache(fullpath)
+        except FileNotFoundError:
+            return False
+
+        if entries is None:
+            return None
+
+        if not self.version_aware or version_id is None:
+            return True
+
+        for entry in entries:
+            if entry["name"] == fullpath and entry.get("VersionId") == version_id:
+                return True
+
+        return False
+
     async def _exists(self, path):
         if path in ["", "/"]:
             # the root always exists, even if anon
             return True
         path = self._strip_protocol(path)
         bucket, key, version_id = self.split_path(path)
         if key:
-            try:
-                if self._ls_from_cache(path):
-                    return True
-            except FileNotFoundError:
-                return False
+            cached_exists = self._exists_from_cache(path, bucket, key, version_id)
+            if cached_exists is not None:
+                return cached_exists
+
```

Comment:
Ok, so I think adlfs is unreasonably falling back to a real API call in https://github.com/fsspec/adlfs/blob/cdd9513013172ac379bf466abaf10e31fc9da6e9/adlfs/spec.py#L1388 . But at least that works 🙂 In s3fs we'll keep closer to old behaviour and will trust dircache by fixing how we interpret it here for versioned paths (only file could versioned, so we try to match version_id and if we fail to do that we report that file doesn't exist in cache).